node-graph/interpreted-executor/src/node_registry.rs

```diff
@@ -444,18 +445,59 @@ fn node_registry() -> HashMap<NodeIdentifier, HashMap<NodeIOTypes, NodeConstruct
 			params: [WasmSurfaceHandleFrame]
 		),
 		async_node!(graphene_core::memo::EndLetNode<_>, input: WasmEditorApi, output: SurfaceFrame, params: [SurfaceFrame]),
-		vec![(
-			NodeIdentifier::new("graphene_core::memo::RefNode<_, _>"),
-			|args| {
-				Box::pin(async move {
-					let node: DowncastBothNode<Option<WasmEditorApi>, WasmEditorApi> = graphene_std::any::DowncastBothNode::new(args[0].clone());
-					let node = <graphene_core::memo::RefNode<_, _>>::new(node);
-					let any: DynAnyNode<(), _, _> = graphene_std::any::DynAnyNode::new(graphene_core::value::ValueNode::new(node));
-					any.into_type_erased()
-				})
-			},
-			NodeIOTypes::new(concrete!(()), concrete!(WasmEditorApi), vec![fn_type!(Option<WasmEditorApi>, WasmEditorApi)]),
-		)],
+		vec![
+			(
+				NodeIdentifier::new("graphene_core::memo::RefNode<_, _>"),
+				|args| {
+					Box::pin(async move {
+						let node: DowncastBothNode<Option<WasmEditorApi>, WasmEditorApi> = graphene_std::any::DowncastBothNode::new(args[0].clone());
+						let node = <graphene_core::memo::RefNode<_, _>>::new(node);
+						let any: DynAnyNode<(), _, _> = graphene_std::any::DynAnyNode::new(graphene_core::value::ValueNode::new(node));
+						any.into_type_erased()
+					})
+				},
+				NodeIOTypes::new(concrete!(()), concrete!(WasmEditorApi), vec![fn_type!(Option<WasmEditorApi>, WasmEditorApi)]),
+			),
+			(
+				NodeIdentifier::new("graphene_std::raster::ImaginateNode<_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _>"),
+				|args: Vec<Arc<graph_craft::proto::NodeContainer>>| {
+					Box::pin(async move {
+						use graphene_std::raster::ImaginateNode;
+						let cache: ImaginateCache = graphene_std::any::input_node(args.last().unwrap().clone()).eval(()).await;
+						macro_rules! instanciate_imaginate_node {
+							($($i:expr,)*) => { ImaginateNode::new($(graphene_std::any::input_node(args[$i].clone()),)* cache.into_inner()) };
+						}
+						let node: ImaginateNode<Color, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _> = instanciate_imaginate_node!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,);
+						let any = graphene_std::any::DynAnyNode::new(ValueNode::new(node));
+						any.into_type_erased()
+					})
+				},
+				NodeIOTypes::new(
+					concrete!(ImageFrame<Color>),
+					concrete!(ImageFrame<Color>),
+					vec![
+						fn_type!(WasmEditorApi),
+						fn_type!(ImaginateController),
+						fn_type!(f64),
+						fn_type!(Option<DVec2>),
+						fn_type!(u32),
+						fn_type!(ImaginateSamplingMethod),
+						fn_type!(f64),
+						fn_type!(String),
+						fn_type!(String),
+						fn_type!(bool),
+						fn_type!(f64),
+						fn_type!(Option<Vec<u64>>),
+						fn_type!(bool),
+						fn_type!(f64),
+						fn_type!(ImaginateMaskStartingFill),
+						fn_type!(bool),
+						fn_type!(bool),
+						fn_type!(ImaginateCache),
+					],
+				),
+			),
+		],
```

Comment:
could you just use the `async_node` macro here?

Reply:
I could, after I figure out how to do the `ImaginateCache`-thing properly. Currently it seems that different nodes share the same cache what obviously isn't desired

Reply:
nodes are considered equal if they receive the same inputs and have the same implementation. 

> We generate a stable node id for each node which you think of sort of like like a Gödel number for a turing machine. It is a unique identifier which describes the mapping from input -> output
the following always has to hold true
let x, y ∈ StableNodeIds
>
> x = y ⇒ ∀i ∈ Input: x.eval(i) == y.eval(i)

So it feels like it would make more sense to store the cache in the node itself anyway instead of feeding it in as an input